### ISSUE TITLE
[java] DoNotUseThreads should not warn on Runnable

### ISSUE THREAD

**Issue by DanySK** [pmd/pmd#1627]
**Affects PMD Version:** 

6.11.0

**Rule:**

[DoNotUseThreads](https://pmd.github.io/pmd-6.11.0/pmd_rules_java_multithreading.html#donotusethreads)

**Description:**

I argue that, since Java 8, Runnables are the de-facto standard way to model `()->void` functions. I am aware that the Javadoc says

>The Runnable interface should be implemented by any class whose instances are intended to be executed by a thread

yet, there is no substitute in `java.util.function`, and until something else is in place, in several cases a delayed computation is modelled through `Runnable`, generating a number of false positives.

I'd like to open a discussion on whether the rule should be changed to take into account only `Thread` (and, possibily, `Executors` and `ExecutorService`).

**Code Sample demonstrating the issue:**

Snippet from Protelis/Protelis

```java
    @Override
    public void eval(final ExecutionContext context) {
        final AnnotatedTree<?> initBranch = getBranch(INIT);
        initBranch.evalInNewStackFrame(context, INIT);
        final S localValue = ensureType(isErased() ? initBranch.getAnnotation() : getSuperscript());
        final AnnotatedTree<?> body = getBranch(BODY);
        ifPresent(localName, it -> context.putVariable(it, localValue, true));
        ifPresent(fieldName, it -> context.putVariable(it, context.buildFieldDeferred(i -> i, localValue, body::getAnnotation), true));
        context.newCallStackFrame(BODY);
        final Runnable yieldEvaluation = () -> ifPresent(yield, it -> it.evalInNewStackFrame(context, YIELD)); // Warning, but this is just a delayed computation, no threads involved...
        if (body instanceof All) {
            final All multilineBody = (All) body;
            multilineBody.forEachWithIndex((i, b) -> {
                context.newCallStackFrame(i.byteValue());
                b.eval(context);
                multilineBody.setAnnotation(b.getAnnotation());
            });
            yieldEvaluation.run();
            multilineBody.forEach(it -> context.returnFromCallFrame());
        } else {
            body.eval(context);
            yieldEvaluation.run();
        }
        context.returnFromCallFrame();
        final S result = ensureType(body.getAnnotation());
        setSuperscript(result);
        setAnnotation(yield.isPresent() ? yield.get().getAnnotation() : (T) result);
    }
```

**Running PMD through:** *[CLI | Ant | Maven | Gradle | Designer | Other]*

**Comment by Vampire** [comment]
Even before it was usable for some design patterns.
I second removing the forbidding of `Runnable`.
Actually even creating an instance of a `Thread` and calling `run()` does not necessirly have to be a violation, as in that case it is not more than a runnable that is run.
Only starting a thread with start is evil.
Actually the rule should instead forbid other things like using exeutor services that use thread pools and so on.

**Comment by DanySK** [comment]
Calling `run()` o a thread is not a violation, but then why using `Thread` (and all its machinery) in place of `Runnable`? It might be indication of a mistake (confusing `run` with `start` is common), so I'd keep that in place for `Thread`, a I see no case in which such a call is made but there is no mistake at all in the code.

**Comment by Vampire** [comment]
`and all its machinery`?
If you only call `run` there is not much machinery.
There are valid reasons to do it, like calling `run` if it should be synchronous but `start` if it should be asynchronous and so on.

But anyway, for the case where `run` was confused with `start`, there is a separate rule that cares about it called `DontCallThreadRun`.

**Comment by DanySK** [comment]
We are going off topic, but still, which is the case in which you should use `Thread` if you need to run synchronously? I can see no case in which you call `Thread.run()` and you could not have used `Runnable`, which is a better choice.
It does not mean they can't exist, I'm arguing they are corner cases, and often they are the result of some design flaw, while `Runnable.run()` is totally legit especially after Java 8.

Anyway, as you say, there is another rule for that specific case.

**Comment by jsotuyod** [comment]
@DanySK @Vampire agreed. The spirit of the rule is to honor J2EE managed approach to threads, so `Runnable` should be ok, and `Thread`, `ThreadGroup`, `Executors`, `ExecutorService` and so forth should be banned.

The `run` / `start` mixup is already addressed in `DontCallThreadRun`. 

This should make an easy fix for anyone looking to contribute, as this is a very basic XPath rule (which could use being revamped to use `typeIs` instead of image matching).

**Comment by mclay** [comment]
@jsotuyod please assign me the issue i will fix it.

**Comment by adangel** [comment]
Awesome, thanks!

**Comment by adangel** [comment]
FYI - I've removed the addition "on java8+" in the issue title, since I don't see this related to java8. Runnable per se doesn't create a new thread regardless of the java version.
